File: src/test/java/com/naminhyeok/fantazzk/ApiResponseTest.java
package com.naminhyeok.fantazzk;

import static org.assertj.core.api.Assertions.assertThat;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;
import org.springframework.http.HttpStatus;
import org.slf4j.event.Level;

class ApiResponseTest {
    private final ObjectMapper objectMapper = new ObjectMapper();

    @Test
    void 성공_응답은_SUCCESS와_payload를_공통_envelope에_담는다() throws Exception {
        ApiResponse<String> response = ApiResponse.success("ok");
        JsonNode json = objectMapper.readTree(objectMapper.writeValueAsString(response));

        assertThat(response.getResultType()).isEqualTo(ResultType.SUCCESS);
        assertThat(response.getSuccess()).isEqualTo("ok");
        assertThat(response.getError()).isNull();
        assertThat(json.get("resultType").asText()).isEqualTo("SUCCESS");
        assertThat(json.get("success").asText()).isEqualTo("ok");
        assertThat(json.get("error").isNull()).isTrue();
    }

    @Test
    void 실패_응답은_descriptor_기반_error_envelope을_사용한다() throws Exception {
        ApiResponse<Void> response = ApiResponse.error(new FakeErrorDescriptor(), "detail");
        JsonNode json = objectMapper.readTree(objectMapper.writeValueAsString(response));

        assertThat(response.getResultType()).isEqualTo(ResultType.ERROR);
        assertThat(response.getSuccess()).isNull();
        assertThat(response.getError().code()).isEqualTo("BAD_REQUEST");
        assertThat(response.getError().message()).isEqualTo("잘못된 요청입니다");
        assertThat(response.getError().data()).isEqualTo("detail");
        assertThat(json.get("resultType").asText()).isEqualTo("ERROR");
        assertThat(json.get("success").isNull()).isTrue();
        assertThat(json.get("error").get("code").asText()).isEqualTo("BAD_REQUEST");
        assertThat(json.get("error").get("message").asText()).isEqualTo("잘못된 요청입니다");
        assertThat(json.get("error").get("data").asText()).isEqualTo("detail");
    }

    private static final class FakeErrorDescriptor implements ErrorDescriptor {

        @Override
        public HttpStatus getStatus() {
            return HttpStatus.BAD_REQUEST;
        }

        @Override
        public String getCode() {
            return "BAD_REQUEST";
        }

        @Override
        public String getMessage() {
            return "잘못된 요청입니다";
        }

        @Override
        public Level getLogLevel() {
            return Level.WARN;
        }
    }
}

File: room/api/src/test/kotlin/com/naminhyeok/fantazzk/room/dto/ApiResponseTest.kt
package com.naminhyeok.fantazzk.room.dto

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test

class ApiResponseTest {
    @Nested
    inner class `성공 응답` {
        @Test
        fun `데이터를 감싸서 SUCCESS 타입으로 반환한다`() {
            val cut = ApiResponse.success("hello")

            assertThat(cut.resultType).isEqualTo(ApiResponse.ResultType.SUCCESS)
            assertThat(cut.success).isEqualTo("hello")
            assertThat(cut.error).isNull()
        }
    }

    @Nested
    inner class `에러 응답` {
        @Test
        fun `에러 정보를 담아 ERROR 타입으로 반환한다`() {
            val cut = ApiResponse.error(404, "NOT_FOUND", "찾을 수 없습니다")

            assertThat(cut.resultType).isEqualTo(ApiResponse.ResultType.ERROR)
            assertThat(cut.success as Any?).isNull()
            assertThat(cut.error).isNotNull()
            assertThat(cut.error!!.status).isEqualTo(404)
            assertThat(cut.error!!.errorCode).isEqualTo("NOT_FOUND")
            assertThat(cut.error!!.reason).isEqualTo("찾을 수 없습니다")
            assertThat(cut.error!!.data).isNull()
        }

        @Test
        fun `에러에 추가 데이터를 포함할 수 있다`() {
            val data = mapOf("field" to "값이 필요합니다")
            val cut = ApiResponse.error(400, "VALIDATION_ERROR", "검증 실패", data)

            assertThat(cut.error!!.data).isEqualTo(data)
        }
    }
}
